Made flag for debugging build time of user created widgets

gaaclarke opened the pull request:
Turning on widget creation tracking for profile builds allows us to do performance evaluation techniques that wouldn't scale across all widgets.  In this case we are using it to track build time for widgets on the CPU but it will also allow us to start tracking raster performance of widgets (cc @iskakaushik)

Added overhead (profile builds):
 * memory
   * ~50 bytes per widget for the construction location (I don't believe we are taking advantage of `const` to reduce this)
   * Memoization of `_isLocalCreationLocation`, ~5 bytes per source file used in the app that constructs widgets.
 * cpu
   *  An extra timeline event for each user created widget's `build`
   * Extra parameter for each widget constructor call
   * String comparisons against construction file uri (memoized with HashMap) for build of each widget

fixes https://github.com/flutter/flutter/issues/100928

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

gaaclarke committed:
Added a bool that allows us to limit debugProfileBuildsEnabled to user
created widgets.

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -742,6 +742,8 @@ mixin WidgetInspectorService {
 
   List<String>? _pubRootDirectories;
+  /// Memoization for [_isLocalCreationLocation].
+  final Map<String, bool> _isLocalCreationCache = <String, bool>{};
```
Faster would be an identity hash from 
Map<Uri, bool>
That will be significantly faster and as these objects are all const, you can trust identity hashes to be safe to use.

gaaclarke replied:
This would require changing the kernel transform to write Uri's into the _Location fields.  We wouldn't want to parse the Uri in order to look up the cached result.  That would potentially wipe out our wins from the identityHash.  Since this is a complicated change, the kernel transform doesn't have any tests, and the performance written this way isn't bad, I'll leave that as a future optimization.

https://github.com/dart-lang/sdk/issues/48694

jacob314 replied:
A couple notes:
there are multiple tests in widget_inspector_test.dart that E2E test the kernel transformer. Adding some more tests there would be good. I'll add one to ensure the _Location fields are still const. If they are not, that is a serious bug.
Actually I was confused. The URI class isn't const so that wouldn't work as the _Location objects need to be const.

gaaclarke commented:
I measured the BUILD times of the Flutter Gallery app to make sure we aren't introducing a lot of overhead by turning on track-widget-creation on in profile builds.  The Flutter Gallery has 3 BUILD's before the first LAYOUT.  Here are the results:

| type | 1st BUILD | 2nd BUILD | 3rd BUILD |
|-|-|-|-|
| w/ track-widget-creation | 6.3 ms | 1.8 ms | 2.1 ms |
| w/o track-widget-creation | 6ms | 1.8 ms | 2.8 ms |

This was tested on an iOS device.  At first blush there doesn't appear to be a significant change in BUILD times as a result of tracking widget creation.

gaaclarke commented:
I ran a performance test with `debugProfileBuildsEnabledUserWidgets` turn off an on as it is currently written:
|debugProfileBuildsEnabledUserWidgets|1st BUILD|2nd BUILD|3rd BUILD|
|-|-|-|-|
|off|6.3 ms|1.8 ms|2.1 ms|
|on|6.5 ms|1.5 ms|2.7 ms|

gaaclarke committed:
made it turned on by default

gaaclarke committed:
switched to hashmap

gaaclarke committed:
Cleaned everything up and added tests

gaaclarke commented:
@Hixie I chose you to review this since you've done work on the Timeline events and we talked about this feature in the past.

gaaclarke committed:
fixed an odd test where it wants to be able to add asserts and run in profile mode

gaaclarke commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -2640,10 +2641,13 @@ class BuildOwner {
           Map<String, String> debugTimelineArguments = timelineArgumentsIndicatingLandmarkEvent;
           assert(() {
-            debugTimelineArguments = element.widget.toDiagnosticsNode().toTimelineArguments();
+            if (kDebugMode) {
```
I know it is weird to check for `kDebugMode` here but there is an explicit test that wants to run profile builds with asserts enabled so this has to be done.  Removing this is technically a breaking change.

Hixie replied:
i don't understand, this ran in profile-mode-with-asserts before, why can't it run in profile-mode-with-asserts now?

gaaclarke replied:
We've explicitly put an exception to block it from working: https://github.com/flutter/flutter/blob/fb60ae53e553486559e1ae0f400c322825808651/packages/flutter/lib/src/foundation/diagnostics.dart#L1570:L1570

That exception states that you shouldn't do this for profile builds.  Jacob was also worried when I showed him that the DiagnosticsNode was being used.

The reason it worked in the past I believe is because there is an early out mechanism higher in toDiagnosticsNode for when you aren't tracking widget creation.

dnfield replied:
Why not just delete the assert and replace it with `if (kDebugMode)`?

gaaclarke replied:
> Why not just delete the assert and replace it with `if (kDebugMode)`?

I didn't do that because it wouldn't functionally be the same since someone can run a debug build with the asserts removed.

dnfield replied:
I think we've really tried to discourage customers from running profile with asserts or debug without. It also seems strange to guard this with "debug + asserts" rather than just debug or just asserts. But it'd be fine to try to tackle this problem more wholistically rather than hanging up this PR on it.

Hixie commented on dev/tracing_tests/test/timeline_test.dart:
```diff
@@ -63,6 +63,7 @@ void main() {
   test('Timeline', () async {
     // We don't have expectations around the first frame because there's a race around
     // the warm-up frame that we don't want to get involved in here.
+    debugProfileBuildsEnabledUserWidgets = false;
```
this ends up making the comment just before this line feel out of place, maybe it should go before the comment?

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/widgets/debug.dart:
```diff
@@ -114,8 +114,17 @@ bool debugPrintGlobalKeyedWidgetLifecycle = false;
+///    [Widget] build.
 bool debugProfileBuildsEnabled = false;
 
+/// Adds [Timeline] events for every user created [Widget] built.
```
typo, should be "user-created"

this should define "user-created".

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/widgets/debug.dart:
```diff
@@ -114,8 +114,17 @@ bool debugPrintGlobalKeyedWidgetLifecycle = false;
+/// Adds [Timeline] events for every user created [Widget] built.
+///
+/// See also:
+///  * [debugProfileBuildsEnabled], functions similarly but shows events for
```
you'll need a blank line after the "See also:"

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/widgets/debug.dart:
```diff
@@ -114,8 +114,17 @@ bool debugPrintGlobalKeyedWidgetLifecycle = false;
+/// Adds [Timeline] events for every user created [Widget] built.
+///
+/// See also:
+///  * [debugProfileBuildsEnabled], functions similarly but shows events for
```
"which functions..."

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -3078,6 +3082,11 @@ class _NotificationNode {
+bool _isProfileBuildsEnabledFor(Widget widget) {
+  return debugProfileBuildsEnabled || (debugProfileBuildsEnabledUserWidgets
+      && debugIsWidgetLocalCreation(widget));
+}
```
the indentation doesn't match the scoping, this might be clearer as:
```dart
  return debugProfileBuildsEnabled
      || (debugProfileBuildsEnabledUserWidgets && debugIsWidgetLocalCreation(widget));
```

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -3072,14 +3084,24 @@ bool debugIsLocalCreationLocation(Object object) {
+/// Returns true if a [Widget] is user created.
+///
+/// This is a faster variant of `debugIsLocalCreationLocation` that is available
+/// in debug and profile builds.
```
the docs should explain why we have `debugIsLocalCreationLocation` (right now the docs suggest this version is faster and available more places, which seems net better in every way)

dnfield replied:
Can we also clarify whether user created means that the user has defined this widget, or whether it means that user code explicitly instantiated this widget?

i.e.

```
// file: main.dart

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Container();
}
```

Is MyWidget user created? Is the container in there user created?

gaaclarke replied:
> the docs should explain why we have debugIsLocalCreationLocation (right now the docs suggest this version is faster and available more places, which seems net better in every way)

Done, `debugIsLocalCreationLocation` takes Object as a parameter and could technically be expanded in the future.  `debugIsWidgetLocalCreationLocation` is just for Widget.

gaaclarke replied:
> Can we also clarify whether user created means that the user has defined this widget, or whether it means that user code explicitly instantiated this widget?

Done, defined in the docstring for the flag.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -3092,14 +3114,18 @@ String? _describeCreationLocation(Object object) {
 ///
 /// Currently creation locations are only available for [Widget] and [Element].
 _Location? _getCreationLocation(Object? object) {
   final Object? candidate =  object is Element && !object.debugIsDefunct ? object.widget : object;
```
stray double-space after equals

gaaclarke replied:
done

Hixie commented on packages/flutter_tools/lib/src/build_system/targets/common.dart:
```diff
@@ -222,7 +222,7 @@ class KernelSnapshot extends Target {
       aot: buildMode.isPrecompiled,
       buildMode: buildMode,
-      trackWidgetCreation: trackWidgetCreation && buildMode == BuildMode.debug,
+      trackWidgetCreation: trackWidgetCreation && buildMode != BuildMode.release,
```
what performance impact does this have?

it changes the behaviour of constant widgets, which seems significant, for one

dnfield replied:
Please do not land this until we've answered this question. Ideally, we'd know how much overhead _just_ this line adds, compared to the rest of this patch, as measured on a low end Android device with a large application like customer: money.

gaaclarke replied:
> it changes the behaviour of constant widgets, which seems significant, for one

Good point.  As far as the Gallery is concerned, the change was negligible in local testing.  I'm open to suggestions to answering this question more definitively.  I can write a microbenchmark but what would be a reasonable way to exercise this?  A widget tree of 1000 const Widgets?

If this is a problem we can tweak track-widget-creation to store its data outside of the Widget.

jacob314 replied:
The change to const behavior would be much smaller if you stored a single bool with each widget indicating whether it is created from the local object or not. That way the only changes in const behavior would be cases where the same exact const widget instance was created by two packages one that happened to be local and one that didn't.

You can't track widget creation data outside the widgets and have it work for const widgets. You could track it outside the widgets for non-const widgets but the performance impact would be significantly higher due to the extra expando/hashmap lookup.

Hixie commented on packages/flutter/lib/src/widgets/debug.dart:
```diff
@@ -114,8 +114,17 @@ bool debugPrintGlobalKeyedWidgetLifecycle = false;
+/// See also:
+///  * [debugProfileBuildsEnabled], functions similarly but shows events for
+///    every widget and has a higher overhead cost.
+bool debugProfileBuildsEnabledUserWidgets = true;
```
this should be added to `debugAssertAllWidgetVarsUnset`

dnfield replied:
Can we not default this to true?

Alternatively, if we do default this true, please provide some numbers on how this impacts internal benchmarks we are tracking. The idea is we should know exactly how much this regresses/adds noise to those benchmarks.

gaaclarke replied:
> this should be added to debugAssertAllWidgetVarsUnset

Done

gaaclarke replied:
> Can we not default this to true?\
\
Alternatively, if we do default this true, please provide some numbers on how this impacts internal benchmarks we are tracking. The idea is we should know exactly how much this regresses/adds noise to those benchmarks.

I added benchmarks.  The discussion about defaulting this to false is below.  We'll see if we can get it a bit faster.  There is Jacob's idea but that is a pretty big change too.

Hixie commented on packages/flutter/lib/src/widgets/debug.dart:
```diff
@@ -114,8 +114,17 @@ bool debugPrintGlobalKeyedWidgetLifecycle = false;
+///    [Widget] build.
 bool debugProfileBuildsEnabled = false;
 
+/// Adds [Timeline] events for every user created [Widget] built.
```
this should mention the default value (especially because it's true)

gaaclarke replied:
The default value shows up in the docs at the bottom of the page where the declaration is shown.

dnfield commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -2640,10 +2641,13 @@ class BuildOwner {
           return true;
         }());
-        if (!kReleaseMode && debugProfileBuildsEnabled) {
+        final bool isTimelineTracked = !kReleaseMode && _isProfileBuildsEnabledFor(element.widget);
```
What's the cost of this?

gaaclarke replied:
All the costs are listed in the description and performance measurements are in the comments.  The TLDR is that with it for the Flutter Gallery executing this versus not sometimes measures faster.  So the added overhead is below the variance.

dnfield replied:
Please check this patch on the internal mobile harness testing on a low end device. Flutter Gallery even on a Moto G4 is not representative, particularly for things that will do extra work on every frame, compared to the internal low end Android device for customer: money.

gaaclarke replied:
responded below.

dnfield commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -3765,10 +3777,13 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
       Map<String, String> debugTimelineArguments = timelineArgumentsIndicatingLandmarkEvent;
       assert(() {
-        debugTimelineArguments = newWidget.toDiagnosticsNode().toTimelineArguments();
+        if (kDebugMode) {
```
ditto

gaaclarke replied:
resolved above

gaaclarke commented:
@dnfield some of the comments were similar so I'll respond to them here.

> Can we not default this to true? \
\
Alternatively, if we do default this true, please provide some numbers on how this impacts internal benchmarks we are tracking. The idea is we should know exactly how much this regresses/adds noise to those benchmarks.

> Please check this patch on the internal mobile harness testing on a low end device. Flutter Gallery even on a Moto G4 is not representative, particularly for things that will do extra work on every frame, compared to the internal low end Android device for customer: money.

> Please do not land this until we've answered this question. Ideally, we'd know how much overhead just this line adds, compared to the rest of this patch, as measured on a low end Android device with a large application like customer: money.

The patch was originally written with the flag set to `false`.  The performance of the the patch was encouraging and lead me to believe the utility outweighs any cost.  Here's some more datapoints that support that decision:
1) We already create a comparable number of Timeline events in the LAYOUT phase.
1) We know from the Timeline benchmarks that Timeline events without parameters are much faster since they aren't doing JSON encoding.  So we have gained cycles there.
1) Any added overhead only effects Profile builds.  If someone is running a Profile build, presumably they want information about the performance of the app, so this is information they want and shouldn't be hidden behind a flag (assuming the cost is reasonable).
1) We know from the Timeline benchmark that it's actually much slower on iOS devices, so testing on iOS is more informative.

Any change to the Framework can adversely affect performance, the bar for landing Framework PR's isn't proving it doesn't have any effect.  The calculus is (user value / performance overhead).  Because this change is fundamental to the Framework I've done due diligence by explaining all performance differences and measuring them locally.

If you have a way I can test customer:money preemptively I'm happy to to do that.  I'm assuming I can apply the patch to a workspace and run a script to get the results from a remote device?  Give me details on what you'd like to test precisely and what results would be acceptable.  If the bar is zero cost to profile builds, that might not be the case, that depends on the savings we get from not doing json encoding in the Timeline event.  If the bar is less than a millisecond for the meanest reasonable scene we can throw at it, I think thats possible.

Maybe a better check would be a microbenchmark with N user widgets?

gaaclarke commented:
@Hixie @dnfield it turns out we already have a good microbenchmark on build times, `flutter/dev/benchmarks/microbenchmarks/lib/stocks/build_bench.dart`

That gives better results than the one I listed above:
|test| result|
|-|-|
| before change | 831.9 µs |
| after change (flag turned off) | 826.1 µs |
| after change (flag turned on) | 969.4 µs |

That shows that the addition of tracking the widget creation and removing the diagnostic nodes in profile builds is a net positive.  The diagnostic nodes had no meaningful data and were just causing JSON encoding to happen.

The net change with the flag on an iOS device (where it will be higher) was 137.5 µs.  Even if that value is off by an order of magnitude it is an acceptable cost.

If you don't think that test is representative let me know and we can tweak it.

dnfield commented:
> Any change to the Framework can adversely affect performance, the bar for landing Framework PR's isn't proving it doesn't have any effect. 

This particular change is touching an are of the framework that is very performance sensitive, so the bar has to be higher for it. And because it has a high likelyhood of impacting a high priority customer's benchmarks, we really should do more DD up front before having a fire drill after a google3 roll that contains 100 commits.

> The net change with the flag on an iOS device (where it will be higher) was 137.5 µs. Even if that value is off by an order of magnitude it is an acceptable cost.

I do not agree. That represents a >10% regression on a microbenchmark. That seems unacceptably high to me to opt this in. Especially since this feature is really providing performance debugging information that is not essential to running an application or otherwise doing performance testing. I am happy to see that there's a slight improvement with the flag turned off, and I would be more willing to approve this PR if it were defaulted to off even if we don't do the internal testing - since I think it's pretty reasonable to assume it will have a net neutral impact on internal benchmarks in that case.

gaaclarke commented:
> That represents a >10% regression on a microbenchmark.

Right, but I would like to point out that it isn't really a scalar on the build time.  If a widget took 200% longer to build the flag would still cost 137.5 µs.  The important measurement is the absolute added overhead given the most complex scene with the worst setup.  I think we can say that that will be likely be below 1ms.

> I would be more willing to approve this PR if it were defaulted to off even if we don't do the internal testing - since I think it's pretty reasonable to assume it will have a net neutral impact on internal benchmarks in that case.

Okay, I'll start working on cleaning everything up since it sounds like we can go ahead one way or another.

> (And I do really appreciate you getting these numbers, it's a big help)

No problem.

gaaclarke commented:
Also in a different PR I'm tweaking the microbenchmarks to also report the probability of the results.  I'm not sure if that will inform the results but it will be nice to see.

gaaclarke committed:
hixie feedback

gaaclarke committed:
hixie2

gaaclarke commented:
Okay, all the cleanup is done.  I'll try a few different things to see if we can get the performance better, otherwise we can turn this off by default.

gaaclarke commented:
I tweaked `_isLocalCreationLocation` to just randomly return true in order to isolate the cost of that function versus the Timeline cost.  The results showed that the 137 µs overhead is mostly the cost of doing Timeline events, so there is no way we can tweak that logic to get the performance we want, we have to make Timeline events faster.

I also compared the performance of tracing all widgets versus user-created widgets on the stocks microbenchmark to see what kind of savings we get when using this flag.
|test|result|
|-|-|
|all widgets| 3308.9 µs |
|user-created widgets| 972.0 µs |

I'll turn this off by default and I'll look into integrating with the work @iskakaushik is doing to turn this feature on temporarily to give performance information on demand.

gaaclarke committed:
made it default to false

dnfield commented on packages/flutter/lib/src/widgets/debug.dart:
```diff
@@ -114,8 +114,21 @@ bool debugPrintGlobalKeyedWidgetLifecycle = false;
 
+/// Adds [Timeline] events for every user-created [Widget] built.
+///
+/// A user-created [Widget] is any [Widget] that is constructed in the root
+/// library.
```
Can we spell this out a little more?

I'm assuming it means something like this:

```dart
import 'package:flutter/material.dart';
import 'package:foo/foo.dart'; // Exposes a `Foo` widget that creates a bunch of child widgets.

// ...
class MyWidget extends StatelessWidget {
  Widget build(BuildContext context) {
    return Container( // Framework will create some child widgets
      color: Colors.red,
      child: Foo( // Foo will create intermediate widgets between here and Container
        propA: 'Baz',
        child: Container(child: ListTile()), // ListTile creates a _bunch_ of widgets
      ),
    );
  }
}
```

Which widgets will show up when this is turned on and I do something like `runApp(MyWidget())`?

dnfield replied:
I forgot to write what I assume: I assume the timeline would show tracing for `MyWidget`, `Container`, `Foo`, `Cotainer`, `ListTile`, and that's it?

gaaclarke replied:
> I forgot to write what I assume: I assume the timeline would show tracing for MyWidget, Container, Foo, Cotainer, ListTile, and that's it?

Yep, that's correct.  What do you suggest we add now that that is clear?

dnfield replied:
Example code that illustrates this.

gaaclarke replied:
I don't know that will help because we'll be saying something like "All the widgets that aren't actually in the code aren't traced".  The interesting thing about this flag is the omission of the things that aren't present in the source code.

I think it would be better if we could explain in prose the difference between what is written now and what I just clarified to you.

How about this:
```
/// A user-created [Widget] is any [Widget] that is constructed in the root
/// library.  Often [Widget]s contain child [Widget]s that are constructed in
/// libraries (for example, a [TextButton] having a [RichText] child).  Timeline events 
/// for those children will be omitted with this flag.
```

dnfield replied:
I think that prose makes sense. And I realize that `constructed` is meant to literally refer to `calling the constructor`. I still think it would help to clearly and explicitly call out that this will hold true whether the widget is defined in the root library or not.

gaaclarke replied:
Done.

jacob314 commented:
> I ran a performance test with `debugProfileBuildsEnabledUserWidgets` turn off an on as it is currently written:
> 
> debugProfileBuildsEnabledUserWidgets	1st BUILD	2nd BUILD	3rd BUILD
> off	6.3 ms	1.8 ms	2.1 ms
> on	6.5 ms	1.5 ms	2.7 ms
> (Tested with the flutter gallery on an iOS device)



> 

Fyi @bkonyi. It would make a big difference if we had a  version of the Dart tracing API with better performance.

dnfield commented:
Did we isolate how much of that cost is runtimeType.toString?

bkonyi commented:
> Fyi @bkonyi. It would make a big difference if we had a version of the Dart tracing API with better performance.

I'm not sure there's many gains to be made. I don't have all the context for this PR, have we determined the timeline events are performing poorly here?

gaaclarke commented:
> > Fyi @bkonyi. It would make a big difference if we had a version of the Dart tracing API with better performance.
> 
> I'm not sure there's many gains to be made. I don't have all the context for this PR, have we determined the timeline events are performing poorly here?

Yes, it's the Timeline events.  I just did work last month making them faster but on iOS they are more than 2x slower than android: https://flutter-flutter-perf.skia.org/e/?begin=1643949613&end=1648138606&queries=sub_result%3Dtimeline_with_arguments&requestType=0

On iOS the majority of the cost is getting the thread time.  This PR adds a filter function that causes more timeline events.  If you make the filter function cost almost nothing it doesn't affect the added overhead.

bkonyi commented:
> On iOS the majority of the cost is getting the thread time.

Ah, I think I remember now. Unfortunately, I don't think there's any way around that since we do need to collect thread timing information for the profile.

gaaclarke commented:
> Ah, I think I remember now. Unfortunately, I don't think there's any way around that since we do need to collect thread timing information for the profile.

This is a bit of a digression for this PR but have you considered giving approximate thread times?  The interval between samples could a parameter.
```dart
double getThreadTime(clockTime) {
  if (clockTime > _threshold) {
    return approximate(_lastClockTime, _lastThreadTime, clockTime);
  } else {
    _lastClockTime = clockTime;
    _lastThreadTime = getExpensiveThreadTime();
    _threshold = clockTime + kInterval;
    return _lastThreadTime;
  }
}
```

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -1518,14 +1522,11 @@ mixin WidgetInspectorService {
-    }
-    final String file = Uri.parse(location.file).path;
+  bool _isLocalCreationLocationImpl(String locationUri) {
+    final String file = Uri.parse(locationUri).path;
```
it isn't the performance bottleneck but why do we perform the uri parsing before looking up in the _isLocalCreationCache? If we used the full locationUri, we would avoid the URI parsing for the common case where the URIs are already in the map.

gaaclarke replied:
I don't believe we are.  Any code inside of `_isLocalCreationLocationImpl` will only get called if the cache doesn't have the value.  So that includes `Uri.parse`.  See the implementation of `isLocalCreationLocation`.

gaaclarke commented:
I implemented the approximation and set the sampling rate to 2,000 hz.  Then I ran the tests with the flag on and off:

|test| time|
|-|-|
| exact thread time | 969.4 µs |
| approximate thread time at 2000 hz | 893.4 µs |

Before the change we were at `831.9 µs`, so we just added 61 µs, would this be acceptable to you @dnfield?  @bkonyi are you open to approximation if I can give you some data about the error?

Here is the patch to dart:
```diff
diff --git a/runtime/vm/os_macos.cc b/runtime/vm/os_macos.cc
index eed635df0d2..5a4c52dbe79 100644
--- a/runtime/vm/os_macos.cc
+++ b/runtime/vm/os_macos.cc
@@ -105,7 +105,14 @@ int64_t OS::GetCurrentMonotonicMicros() {
   return GetCurrentMonotonicTicks() / kNanosecondsPerMicrosecond;
 }
 
-int64_t OS::GetCurrentThreadCPUMicros() {
+namespace {
+struct ThreadTimeInfo {
+  int64_t threshold = 0;
+  int64_t last_thread_time = 0;
+  int64_t last_clock_time = 0;
+};
+
+int64_t ExpensiveGetCurrentThreadCPUMicros() {
   mach_msg_type_number_t count = THREAD_BASIC_INFO_COUNT;
   thread_basic_info_data_t info_data;
   thread_basic_info_t info = &info_data;
@@ -121,6 +128,28 @@ int64_t OS::GetCurrentThreadCPUMicros() {
   return thread_cpu_micros;
 }
 
+int64_t approximate(const ThreadTimeInfo* info, int64_t clock_time) {
+  return (clock_time - info->last_clock_time) + info->last_thread_time;
+}
+
+thread_local ThreadTimeInfo* s_info = nullptr;
+} // namespace
+
+int64_t OS::GetCurrentThreadCPUMicros() {
+  int64_t clock_time = GetCurrentTimeMicros();
+  if (s_info == nullptr) {
+    s_info = new ThreadTimeInfo();
+  }
+  if (clock_time > s_info->threshold) {
+    s_info->last_clock_time = clock_time;
+    s_info->last_thread_time = ExpensiveGetCurrentThreadCPUMicros();
+    s_info->threshold = clock_time + 500;
+    return s_info->last_thread_time;
+  } else {
+    return approximate(s_info, clock_time);
+  }
+}
+
 int64_t OS::GetCurrentThreadCPUMicrosForTimeline() {
   return OS::GetCurrentThreadCPUMicros();
 }
```

gaaclarke commented:
@dnfield nevermind the approximation question, let's land this turned off and I can follow up later about turning it on with improvements to Timeline in a different PR.

gaaclarke committed:
updated docstring as per dans request

bkonyi commented:
> I implemented the approximation and set the sampling rate to 2,000 hz. Then I ran the stocks microbenchmark with the flag on comparing it with baseline dart:
> 
> test	time
> exact thread time	969.4 µs
> approximate thread time at 2000 hz	893.4 µs
> Before the change we were at `831.9 µs`, so we just added 61 µs, would this be acceptable to you @dnfield? @bkonyi are you open to approximation if I can give you some data about the error?
> 
> Here is the patch to dart:
> ...

@rmacnak-google might have a better way of doing this that doesn't involve approximating thread time and is quite a bit faster (quick results show ~3x improvement on some simple C programs when using the current `GetCurrentThreadCPUMicros()` implementation and one that uses `clock_gettime_nsec_np(CLOCK_THREAD_CPUTIME_ID) / 1000` instead). I'll follow up later whether or not this approach pans out.

jacob314 reviewed (approved):
![lgtm](https://i.imgur.com/JgNRalI.png)

gaaclarke commented:
I just tested `build_bench.dart` with this flag on with the 2 new patches we've done to make Timeline events faster.
flag off: 789.5
flag on: 930.6

It didn't lower the percentage of overhead.  It just dropped the overall numbers.  I'm going to walk away from this now that the low hanging fruit has been explored.
